pyquil/gates.py

```diff
@@ -14,19 +14,30 @@
 #    limitations under the License.
 ##############################################################################
 from warnings import warn
-from typing import Callable, Dict
-from pyquil.quilatom import unpack_qubit, unpack_classical_reg, MemoryReference, Addr, Qubit
-from pyquil.quilbase import (Measurement, Gate, Wait, Reset, Halt, Nop,
+from typing import Any, Callable, Dict, Optional, Tuple, Union
+
+from pyquil.quilatom import (Addr, MemoryReference, MemoryReferenceDesignator, Parameter, Qubit,
+                             QubitDesignator, QubitPlaceholder, unpack_classical_reg, unpack_qubit)
+from pyquil.quilbase import (AbstractInstruction, Gate, Halt, Reset, ResetQubit, Measurement, Nop,
+                             Wait,
                              ClassicalNeg, ClassicalNot,
                              ClassicalAnd, ClassicalInclusiveOr, ClassicalExclusiveOr,
                              ClassicalEqual, ClassicalGreaterEqual, ClassicalGreaterThan,
                              ClassicalLessEqual, ClassicalLessThan,
                              ClassicalMove, ClassicalExchange, ClassicalConvert,
                              ClassicalLoad, ClassicalStore,
-                             ClassicalAdd, ClassicalSub, ClassicalMul, ClassicalDiv, ResetQubit)
+                             ClassicalAdd, ClassicalSub, ClassicalMul, ClassicalDiv)
+
+
+MemoryReferenceOrImmediateValue = Union[MemoryReferenceDesignator, int, float]
+MemoryReferenceOrImmediateInt = Union[MemoryReferenceDesignator, int]
+# TODO:(appleby) Should this type just be quilatom.ParameterDesignator?
```

Comment:
It looks like the answer is "yes", since arbitrary `Expression`s are allowed in place of `GateParameter`s  more restrictive `Parameter` type specifier here.